### ISSUE TITLE
xgboost crash in multi-class prediction

### ISSUE THREAD

**Issue by deanmckee007** [dmlc/xgboost#207]
I create models in parallel using 'foreach' and 'doParallel'.  The models themselves are stored in a list.  I also get the predictions in parallel, but I've displayed the code below for just predicting with the first model in the list.  To replicate, `trainFull` is just the training data from the Otto Kaggle competition; `test` is the test data; `target` is just a numeric vector of the dependent variable; `modelVars` is just a vector of variable names (without id and the target)

```
library('foreach')
library('doParallel')
library('xgboost')

BootStrappedModels <- function(seed){
    set.seed(seed)
    obsFilter <- sample(1:nrow(trainFull[,]),nrow(trainFull[,]),replace=TRUE)
    targetFilter <- target[]
    tempTarget <- targetFilter[obsFilter]
   
    x <- as.matrix(trainFull[obsFilter,modelVars])
    x <- matrix(as.numeric(x),nrow=nrow(x),ncol=ncol(x))
    tempModel <- xgboost(#param=param, 
                         data=x, 
                         label=(tempTarget)-1,
                         nrounds=250,
    )
    return (tempModel)
}

cl <- makeCluster(5)
registerDoParallel(cl)

models <- foreach (i=(1:5), .packages=c(,'xgboost')) %dopar% {
  BootStrappedModels(i)
}

stopCluster(cl)
x <- as.matrix(test[,modelVars])
x <- matrix(as.numeric(x),nrow=nrow(x),ncol=ncol(x))
preds <- predict(models[[1]],x) ####This is where failure occurs
preds <- matrix(preds,9,length(preds)/9)
preds <- t(preds)
```

**Comment by hetong007** [comment]
Hi @deanmckee007 , can you provide the parameters `param` as well?

**Comment by deanmckee007** [comment]
I'll try running it with default values - I'm not willing to share the actual parameters I use while the competition is running.

**Comment by hetong007** [comment]
@deanmckee007 I understand. I just want to make sure it is reproducible with the same setting as the [sample code](https://github.com/dmlc/xgboost/blob/master/demo/kaggle-otto/otto_train_pred.R).

**Comment by deanmckee007** [comment]
Ok, I'll verify that I still get a crash with those parameters.

**Comment by hetong007** [comment]
@deanmckee007 That's helpful, thanks. I will look into that.

**Comment by deanmckee007** [comment]
Yep, also crashed with the posted parameters

param <- list("objective" = "multi:softprob",
"eval_metric" = "mlogloss",
"num_class" = 9,
"nthread" = 8)

**Comment by hetong007** [comment]
@tqchen It can be reproduced with the following code:

```r
library('foreach')
library('doParallel')
library('xgboost')

x = matrix(rnorm(1:1000),200,5)
target = sample(0:2,200,replace = TRUE)

param <- list("objective" = "multi:softprob",
              "eval_metric" = "mlogloss",
              "num_class" = 3,
              "nthread" = 2)

BootStrappedModels <- function(seed){
    tempModel <- xgboost(params=param, data=x, label=(target), nrounds=5)
    return (tempModel)
}

cl <- makeCluster(2)
registerDoParallel(cl)

models <- foreach (i=(1:2), .packages=c('xgboost')) %dopar% {
  BootStrappedModels(i)
}

stopCluster(cl)
preds <- predict(models[[1]],x) ####This is where failure occurs
```

Use `valgrind` to exam:

```
# valgrind --leak-check=full --log-file='valgrind.log' -v Rscript issue207.R
Loading required package: iterators
Loading required package: parallel

 *** caught segfault ***
address 0x190, cause 'memory not mapped'

Traceback:
 1: .Call("XGBoosterPredict_R", object$handle, newdata, as.integer(option),     as.integer(ntreelimit), PACKAGE = "xgboost")
 2: .local(object, ...)
 3: predict(models[[1]], x)
 4: predict(models[[1]], x)
aborting ...
Segmentation fault (core dumped)
```

And the corresponding log file is 

```
==1624== Memcheck, a memory error detector
==1624== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==1624== Using Valgrind-3.10.0.SVN and LibVEX; rerun with -h for copyright info
==1624== Command: Rscript issue207.R
==1624== Parent PID: 1569
==1624== 
--1624-- 
--1624-- Valgrind options:
--1624--    --leak-check=full
--1624--    --log-file=valgrind.log
--1624--    -v
--1624-- Contents of /proc/version:
--1624--   Linux version 3.13.0-44-generic (buildd@lamiak) (gcc version 4.8.2 (Ubuntu 4.8.2-19ubuntu1) ) #73-Ubuntu SMP Tue Dec 16 00:22:43 UTC 2014
--1624-- Arch and hwcaps: AMD64, amd64-cx16-rdtscp-sse3-avx
--1624-- Page sizes: currently 4096, max supported 4096
--1624-- Valgrind library directory: /usr/lib/valgrind
--1624-- Reading syms from /usr/bin/Rscript
--1624--   Considering /usr/bin/Rscript ..
--1624--   .. CRC mismatch (computed d9ea24ea wanted 67fbe2d6)
--1624--    object doesn't have a symbol table
--1624-- Reading syms from /lib/x86_64-linux-gnu/ld-2.19.so
--1624--   Considering /lib/x86_64-linux-gnu/ld-2.19.so ..
--1624--   .. CRC mismatch (computed 4cbae35e wanted 8d683c31)
--1624--   Considering /usr/lib/debug/lib/x86_64-linux-gnu/ld-2.19.so ..
--1624--   .. CRC is valid
--1624-- Reading syms from /usr/lib/valgrind/memcheck-amd64-linux
--1624--   Considering /usr/lib/valgrind/memcheck-amd64-linux ..
--1624--   .. CRC mismatch (computed 37cdde19 wanted adc367dd)
--1624--    object doesn't have a symbol table
--1624--    object doesn't have a dynamic symbol table
--1624-- Scheduler: using generic scheduler lock implementation.
--1624-- Reading suppressions file: /usr/lib/valgrind/default.supp
==1624== embedded gdbserver: reading from /tmp/vgdb-pipe-from-vgdb-to-1624-by-ubuntu-on-???
==1624== embedded gdbserver: writing to   /tmp/vgdb-pipe-to-vgdb-from-1624-by-ubuntu-on-???
==1624== embedded gdbserver: shared mem   /tmp/vgdb-pipe-shared-mem-vgdb-1624-by-ubuntu-on-???
==1624== 
==1624== TO CONTROL THIS PROCESS USING vgdb (which you probably
==1624== don't want to do, unless you know exactly what you're doing,
==1624== or are doing some strange experiment):
==1624==   /usr/lib/valgrind/../../bin/vgdb --pid=1624 ...command...
==1624== 
==1624== TO DEBUG THIS PROCESS USING GDB: start GDB like this
==1624==   /path/to/gdb Rscript
==1624== and then give GDB the following command
==1624==   target remote | /usr/lib/valgrind/../../bin/vgdb --pid=1624
==1624== --pid is optional if only one valgrind process is running
==1624== 
--1624-- REDIR: 0x4019ca0 (strlen) redirected to 0x38068331 (???)
--1624-- Reading syms from /usr/lib/valgrind/vgpreload_core-amd64-linux.so
--1624--   Considering /usr/lib/valgrind/vgpreload_core-amd64-linux.so ..
--1624--   .. CRC mismatch (computed 329d6860 wanted c0186920)
--1624--    object doesn't have a symbol table
--1624-- Reading syms from /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so
--1624--   Considering /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so ..
--1624--   .. CRC mismatch (computed 1fb85af8 wanted 2e9e3c16)
--1624--    object doesn't have a symbol table
==1624== WARNING: new redirection conflicts with existing -- ignoring it
--1624--     old: 0x04019ca0 (strlen              ) R-> (0000.0) 0x38068331 ???
--1624--     new: 0x04019ca0 (strlen              ) R-> (2007.0) 0x04c2e1a0 strlen
--1624-- REDIR: 0x4019a50 (index) redirected to 0x4c2dd50 (index)
--1624-- REDIR: 0x4019c70 (strcmp) redirected to 0x4c2f2f0 (strcmp)
--1624-- REDIR: 0x401a9c0 (mempcpy) redirected to 0x4c31da0 (mempcpy)
--1624-- Reading syms from /lib/x86_64-linux-gnu/libpthread-2.19.so
--1624--   Considering /lib/x86_64-linux-gnu/libpthread-2.19.so ..
--1624--   .. CRC mismatch (computed d568bf33 wanted fb00d679)
--1624--   Considering /usr/lib/debug/lib/x86_64-linux-gnu/libpthread-2.19.so ..
--1624--   .. CRC is valid
--1624-- Reading syms from /lib/x86_64-linux-gnu/libc-2.19.so
--1624--   Considering /lib/x86_64-linux-gnu/libc-2.19.so ..
--1624--   .. CRC mismatch (computed dc620abc wanted 148cbd6e)
--1624--   Considering /usr/lib/debug/lib/x86_64-linux-gnu/libc-2.19.so ..
--1624--   .. CRC is valid
--1624-- REDIR: 0x50e1d60 (strcasecmp) redirected to 0x4a25720 (_vgnU_ifunc_wrapper)
--1624-- REDIR: 0x50e4050 (strncasecmp) redirected to 0x4a25720 (_vgnU_ifunc_wrapper)
--1624-- REDIR: 0x50e1530 (memcpy@GLIBC_2.2.5) redirected to 0x4a25720 (_vgnU_ifunc_wrapper)
--1624-- REDIR: 0x50df7c0 (rindex) redirected to 0x4c2da30 (rindex)
--1624-- REDIR: 0x50d7750 (malloc) redirected to 0x4c2ab10 (malloc)
--1624-- REDIR: 0x50ddac0 (strlen) redirected to 0x4c2e0e0 (strlen)
--1624-- REDIR: 0x50e8ac0 (strchrnul) redirected to 0x4c319b0 (strchrnul)
--1624-- REDIR: 0x50dc070 (strcmp) redirected to 0x4a25720 (_vgnU_ifunc_wrapper)
--1624-- REDIR: 0x5195200 (__strcmp_ssse3) redirected to 0x4c2f1b0 (strcmp)
--1624-- REDIR: 0x5090910 (putenv) redirected to 0x4c32520 (putenv)
--1624-- REDIR: 0x50dbe50 (__GI_strchr) redirected to 0x4c2db90 (__GI_strchr)
--1624-- REDIR: 0x50ddc80 (strnlen) redirected to 0x4c2e080 (strnlen)
--1624-- REDIR: 0x50e6780 (__GI_memcpy) redirected to 0x4c2fc90 (__GI_memcpy)
--1624-- REDIR: 0x50ddf30 (__GI_strncmp) redirected to 0x4c2e930 (__GI_strncmp)
--1624-- REDIR: 0x50d7ef0 (realloc) redirected to 0x4c2ce10 (realloc)
```

**Comment by tqchen** [comment]
This issue was due to the incomplete check of externelptr to be null. It is now fixed in newest commit. 

Thanks!

**Comment by deanmckee007** [comment]
Fantastic - you guys continue to impress.  Look forward to trying it out again in the next few days.

**Comment by hetong007** [comment]
Confirmed on my side.

**Comment by deanmckee007** [comment]
Guys, although R no longer crashes, another error pops up:

Error in .local(object, ...) : must set num_class to use softmax

You can reproduce by using the code hetong007 posted above.

**Comment by tqchen** [comment]
If you are indeed working on multi-class classification. This was not an error, but just a message to ask you to set  num_class

**Comment by deanmckee007** [comment]
num_class is specified... I'm reposting his code from above:

library('foreach')
library('doParallel')
library('xgboost')

x = matrix(rnorm(1:1000),200,5)
target = sample(0:2,200,replace = TRUE)

param <- list("objective" = "multi:softprob",
              "eval_metric" = "mlogloss",
              "num_class" = 3,
              "nthread" = 2)

BootStrappedModels <- function(seed){
    tempModel <- xgboost(params=param, data=x, label=(target), nrounds=5)
    return (tempModel)
}

cl <- makeCluster(2)
registerDoParallel(cl)

models <- foreach (i=(1:2), .packages=c('xgboost')) %dopar% {
  BootStrappedModels(i)
}

stopCluster(cl)
preds <- predict(models[[1]],x) ####This is where failure occurs

**Comment by hetong007** [comment]
@tqchen I can reproduce this problem with the following code:

```r
require(xgboost)

x = matrix(rnorm(1:1000),200,5)
target = sample(0:2,200,replace = TRUE)

param <- list("objective" = "multi:softprob",
              "eval_metric" = "mlogloss",
              "num_class" = 3,
              "nthread" = 2)

Model <- xgboost(params=param, data=x, label=(target), nrounds=5)
preds <- predict(Model,x)

Raw_info = xgb.save.raw(Model)
Model2 = xgb.load(Raw_info)
preds <- predict(Model2,x)
identical(Model$raw,Model2$raw)
```

Predicting from `Model` is fine while from `Model2` is not good. The model loaded by `xgb.load` seems lack of some parameter informations.

**Comment by hetong007** [comment]
@deanmckee007 Do you mind to update `xgboost` and try again? @tqchen just pushed a fix. It is fine now on my side. Thanks!

**Comment by deanmckee007** [comment]
Run 'devtools::install_github('dmlc/xgboost',subdir='R-package')' to update?

**Comment by hetong007** [comment]
@deanmckee007 yes. Then restart your R to make sure the loaded xgboost is the latest.

**Comment by deanmckee007** [comment]
Finally got a chance to try it out.  Confirmed fixed on my end as well.
